earthfile2llb/converter.go

```diff
@@ -1328,9 +1353,36 @@ func (c *Converter) FinalizeStates(ctx context.Context) (*states.MultiTarget, er
 	return c.mts, nil
 }
 
+// hasLeadingShellOut returns true if s starts with a shell-out
+func hasLeadingShellOut(s string) bool {
+	return strings.HasPrefix(s, "$(") || strings.HasPrefix(s, "\"$(")
+}
+
 // ExpandArgs expands args in the provided word.
-func (c *Converter) ExpandArgs(word string) string {
-	return c.varCollection.Expand(word)
+func (c *Converter) ExpandArgs(ctx context.Context, runOpts ConvertRunOpts, word string) (string, error) {
+	if !c.opt.Features.ShellOutAnywhere {
+		if hasLeadingShellOut(word) {
+			runOpts.Args = []string{word}
+			return c.RunExpression(ctx, "internal-expand-args", runOpts)
+		}
+
+		splitArg := strings.SplitN(word, "=", 2)
+		if len(splitArg) == 2 && hasLeadingShellOut(splitArg[1]) {
+			runOpts.Args = []string{splitArg[1]}
+			expanded, err := c.RunExpression(ctx, "internal-expand-args", runOpts)
+			if err != nil {
+				return "", err
+			}
+			return fmt.Sprintf("%s=%s", splitArg[0], expanded), nil
+		}
+
+		return c.varCollection.ExpandOld(word), nil
+	}
+	s := c.varCollection.Expand(word, func(cmd string) (string, error) {
+		runOpts.Args = []string{cmd}
+		return c.RunCommand(ctx, "internal-expand-args", runOpts)
+	})
+	return s, nil
```

Comment:
That might be tricky, since this function needs to call `c.RunExpression` and `c.RunCommand` (where `c` is `*Converter`).

I could move some of the logic that's required to maintain the old behaviour into `varCollection.ExpandOld` and pass along either a callback to `RunExpression` or a `*Converter` pointer. I'm not sure if it'll clean things up to be honest.

Reply:
> That might be tricky, since this function needs to call c.RunExpression and c.RunCommand (where c is *Converter).

You can always create a method of Converter that returns a closure. That way the function has access to the converter. I don't think we need to pass in the converter pointer itself.

(The old `processNonConstantBuildArgFunc` did exactly that)

Reply:
I restored `processNonConstantBuildArgFunc` and was able to delete the majority of this code -- now we'll use `processNonConstantBuildArgFunc` to maintain backwards compatibility (both here and in other parts of the converter, as we did before). When we have `--shell-out-anywhere` set, we'll process the shell expansion here all in a single-shot.